### PR TITLE
drivers/i2c: stm32: Use dts for i2c clock source

### LINKED ISSUE #52675: I2C: STM32F0 can't switch to HSI clock by default and change timing
Even with the dts correctly set we can't set the HSI clock for I2C1 peripheral and change the timings accordingly.

```
clocks = <&rcc STM32_CLOCK_BUS_APB1 0x00200000>,
	         <&rcc STM32_SRC_HSI I2C1_SEL(2)>;
	clock-frequency = <I2C_BITRATE_STANDARD>;
	timings = <48000000 I2C_BITRATE_STANDARD 0x20303E5D>,
	          <48000000 I2C_BITRATE_FAST 0x2010091A>,
	          <8000000 I2C_BITRATE_STANDARD 0x2000090E>,
	          <8000000 I2C_BITRATE_FAST 0x0000020B>;
```

The clock is instantly set to SYSCLK and it doesn't check if another clock is set inside the clocks property.
https://github.com/zephyrproject-rtos/zephyr/blob/2bfc4451cf4d39b585d9a65f86d8b38facc84277/drivers/i2c/i2c_ll_stm32.c#L238-L250

We can't even use the `i2c_configure` function as it also force the clock to SYSCLK.

Actually, the only way is to disable the I2C, change the clock with `clock_control_configure`, change the timing with `LL_I2C_SetTiming` and enable the I2C inside the application code...

Is it possible to change the driver as to make sure the right clock is selected on init or when we use `clock_control_configure` and `i2c_configure` ?

### PR DESCRIPTION
**Pull request by gautierg-st**
This PR adds support of the I2C clock source for STM32 in the dts file.

Fixes #52675

### CODE REVIEW

File: include/zephyr/dt-bindings/clock/stm32f4_clock.h [hunk 1]
```diff
@@ -68,4 +71,27 @@
 /** BDCR devices */
 #define RTC_SEL(val)		STM32_CLOCK(val, 3, 8, BDCR_REG)
 
+/* Following definitions are for STM32F7X only */
```
Comment (erwango) [inline include/zephyr/dt-bindings/clock/stm32f4_clock.h:74]: Please introduce a dedicated stm32f7_clocks.h

File: dts/arm/st/f0/stm32f0.dtsi [hunk 2]
```diff
@@ -179,7 +179,8 @@
 			#address-cells = <1>;
 			#size-cells = <0>;
 			reg = <0x40005400 0x400>;
-			clocks = <&rcc STM32_CLOCK_BUS_APB1 0x00200000>;
```
Comment (erwango) [inline dts/arm/st/f0/stm32f0.dtsi:182]: Please provide more comments in the commit message why we are defining default clock sources (which is not a common practice). According to existing code, I can see:
```
	 * STM32F0/3 I2C's independent clock source supports only
	 * HSI and SYSCLK, not APB1. We force I2C clock source to SYSCLK.
	 * I2C2 on STM32F0 uses APB1 clock as I2C clock source
```

File: drivers/i2c/i2c_ll_stm32.c [hunk 3]
```diff
@@ -230,39 +230,20 @@ static int i2c_stm32_init(const struct device *dev)
...
-		LL_RCC_SetI2CClockSource(LL_RCC_I2C2_CLKSOURCE_SYSCLK);
-		break;
-#endif
-
-#if DT_NODE_HAS_STATUS(DT_NODELABEL(i2c3), okay)
-	case DT_REG_ADDR(DT_NODELABEL(i2c3)):
-		LL_RCC_SetI2CClockSource(LL_RCC_I2C3_CLKSOURCE_SYSCLK);
-		break;
-#endif
+#if (DT_NUM_CLOCKS(DT_NODELABEL(name)) > 1)
```
Comment (erwango) [inline drivers/i2c/i2c_ll_stm32.c:238]: Contrary to `STM32_I2C_INIT()` case, `name` is undefined, this will likely not produce the expected result.

**Review by erwango — CHANGES_REQUESTED** [review]

File: include/zephyr/dt-bindings/clock/stm32f7_clock.h [hunk 4]
```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) 2022 Linaro Limited
```
Comment (erwango) [inline include/zephyr/dt-bindings/clock/stm32f7_clock.h:2]: Fix copyright

**Review by erwango — CHANGES_REQUESTED** [review]

**Review by erwango — APPROVED** [review]
LGTM

**Review by carlescufi — APPROVED** [review]

**Pull request merged**